### PR TITLE
METRON-1135: Make Stellar field transformations handle nulls properly

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
As it stands, stellar field transformations disallow users to set variables to null (and thus remove the field). This should be corrected. For instance,
```
"fieldTransformations" : [
        { "transformation" : "STELLAR"
        ,"output" : [ "new_field", "new_field2"]
        ,"config" : {
          "new_field" : "old_field"
         ,"new_field2" : "old_field2"
         ,"old_field" : "null"
         ,"old_field2" : "null"
                    }
        }
]
```
should rename "old_field" to "new_field and "old_field2" to "old_field2" The current behavior is to keep all 4 fields and to not change old_field or old_field2 at all.

Manual testing:
* Create a sensor with the field transformation above.
* Send data through with `old_field` set to a value
* Ensure that `old_field` does not exist in any message in the index for your sensor and it is replaced with `new_field`.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by ottobackwards** [comment]
Would it not be better to have different transformers to specify the behavior here?  What is here is correct, but what if the old behavior was fine?  What if they don't want the fields removed?

**Comment by cestella** [comment]
So, the problem with the old way was that it's confusing and just wrong. Perhaps an example would help here.  For the message
```
{
  "title" : "otto is great"
}
```
If I wanted to rename title to `better_title`, I'd expect the following to work:
```
{ 
  "transformation" : "STELLAR"
 ,"output" : [ "better_title", "title"]
 ,"config" : {
          "better_title" : "title"
         ,"title" : "null"
                    }
}
```
At the end of this, I'd expect either `title` to be gone *or* at the very least, for it to be set to `null` as that is what I directed it to do, but in actuality, it'll still be `otto is great`.  In fact, the output would be:
```
{
  "title" : "otto is great",
  "better_title" : "otto is great"
}
```

This becomes even MORE insidious if we start doing things like returning null from Stellar functions.

**Comment by ottobackwards** [comment]
if(outputs.contains(oField)) {
              ret.put(oField, o);
            }
            if( o != null ) {
              intermediateVariables.put(oField, o);
            }

**Comment by cestella** [comment]
@ottobackwards good suggestion.  I've adopted it.

**Comment by ottobackwards** [comment]
Ok, the question here is this.  I have 3 statements in the transform

{
  "title" : "casey is stellar"
}

{ 
  "transformation" : "STELLAR"
 ,"output" : [ "better_title", "title"]
 ,"config" : {
          "better_title" : "title"
         ,"title" : "SOME_UDF_THAT_CAN_RET_NULL()"
         ,"title" : "if title == null then GET_RANDOM_TITLE()
                    }
}

If the second function results in the removal of title... what do we do?  What was the intent of that null ( from a UDF maybe ), we don't know if it was to remove or not.

**Comment by cestella** [comment]
Ok, I think I see the point of confusion.  Here's the issue; in every other place where we use stellar (i.e. enrichments) assignment to null implies a variable removal.  In field transformations, assignment to null is noop.  I'm just trying to make them consistent.  We could make both noop, but then we'd remove the ability to have temporary variables in enrichment, which has proven very useful.  Also, if we keep the current functionality, the example above, I'd argue, yields confusing results.

That being said, I think this is probably the WRONG behavior in both cases.  What I'd propose, instead, is support within stellar to be able to modify the variables.  I'd propose:
* `REMOVE('field')` would remove the field
* `^=` instead of `:=` would create a temporary variable so you would not need to remember to remove.  e.g. `foo ^= TO_UPPER('casey')` would not be persisted to the message in either field transformations or enrichments.

I think the current functionality is a bit kludgey, but I think as a stopgap we should at least make them do the same thing and I think removal is the most sensible choice.

**Comment by cestella** [comment]
Currently you cannot have two statements in stellar field transformations with the same variable, so 
```
{
"transformation" : "STELLAR"
,"output" : [ "better_title", "title"]
,"config" : {
"better_title" : "title"
,"title" : "SOME_UDF_THAT_CAN_RET_NULL()"
,"title" : "if title == null then GET_RANDOM_TITLE()
}
}
```
would evaluate only the last `title` expression.  It's a problem with our use of maps and why in enrichment we transitioned to lists of strings as an alternative 😢

**Comment by cestella** [comment]
This goes to another deficiency, field transformations don't act like enrichment.  YOu have to indicate the output fields to order the statements of evaluation.  This is ridiculous; we should allow an array of stellar expressions which are evaluated and incrementally update the message, like we have in enrichments.

**Comment by ottobackwards** [comment]
Should there be a documentation change for this, like a 

> note about field transformations and null
> 1. sorry
> 2. here is how it works

**Comment by cestella** [comment]
@ottobackwards Definitely there should and I'll make it as part of this PR.

**Comment by ottobackwards** [comment]
ok, I'm +1 pending the doc change.  Thanks for explaining.

**Comment by cestella** [comment]
Ok, I updated the docs with a summary of what we talked about and some discussion of gotchas.  Hope that is sufficient; if not, let me know.

**Comment by ottobackwards** [comment]
+1, thanks for the contribution!

File: metron-platform/metron-parsers/README.md [hunk 1]
```diff
@@ -212,6 +212,49 @@ into `{ "protocol" : "TCP", "source.type" : "bro", ...}`
...
+This would set `new_field` to the value of `old_field` and remove `old_field`.
+
+### Warning: Transforming the same field twice
+
+Currently, the stellar expressions are expressed in the form of a map where the keys define
+the fields and the values define the Stellar expressions.  You order the expression evaluation
+in the `output` field.  A consequence of this choice to store the assignments as a map is that
+the same field cannot appear in the map as a key twice.
+
+For instance, the following will not function as expected:
```
Comment (justinleet) [inline metron-platform/metron-parsers/README.md]: I think this should be reworded. It's not that it doesn't function as expected; that would imply a bug or an improvement that needs to happen.

Comment (justinleet) [reply metron-platform/metron-parsers/README.md]: Sigh, I don't read well, it sounds like this actually doesn't function as expected based on other expectations, and this is fine until we fix it.

**Comment by justinleet** [comment]
Forgive my lack of understanding (possibly remembering?) of some of the underlying Stellar assumptions, but it seems like we mix up the meaning of actual null and String null.  It may not be entirely reasonable, but I would expect "null" to be an actual string null.

For example:
```
          "better_title" : "title"
         ,"title" : "null"
```
There's no formatting or any other distinction between "title" and "null" that would imply null is treated as a special case in string handling.

JSON has an actual null value that's not string null, and it seems like if we're defining transforms in JSON that's how it should be handled, e.g.
```
          "better_title" : "title"
         ,"title" : null
```

**Comment by ottobackwards** [comment]
@justinleet, the value side of the json is a string, that is a Stellar statement.  So you are not passing in a null statement, you are passing in a Stellar statement that evaluates to null.  In this case via the null keyword.

If you wanted to pass a string that said null, you would have to have:
```
,"title" : "'null'"
```

**Comment by ottobackwards** [comment]
```
[Stellar]>>> Functions loaded, you may refer to functions now...

[Stellar]>>> foo := null
[Stellar]>>> foo
[Stellar]>>> foo := 'null'
[Stellar]>>> foo
null
[Stellar]>>>
```

**Comment by justinleet** [comment]
@ottobackwards Thanks for the refresher on that.  I'm +1

**Pull request closed without merging**